scanner/decisions/0009-nvd-cvss-scores.md

```diff
@@ -0,0 +1,185 @@
+# 0009 - NVD CVSS Scores
+
+- **Author(s):** Ross Tannenbaum and Yi Li
+- **Created:** [2024-04-11 Thurs]
+
+## Status
+
+Status: Accepted.
+
+## Context
+
+Users who require FedRAMP compliance [must be able see NVD's CVSS scores associated with each CVE](https://www.fedramp.gov/assets/resources/documents/CSP_Vulnerability_Scanning_Requirements.pdf).
+Note: NVD tracks CVEs, so all of its data is CVE-based.
+
+Currently, Scanner V4 provides a single CVSS score and severity, which is preferably from the vendor.
+In reality, the CVSS scores that are shown are from NVD except in the following cases:
+
+* RHEL-base images
+  * Red Hat-provided CVSS scores/severities are used.
+* OSV-provided data
+  * If OSV.dev has a CVSS score, then that score is used.
+    * The severity is derived from the score based on the CVSS version's specification.
+  * Note: From personal experience, it is very common for the score displayed by OSV is the same as NVD.
+
+Currently, there is only a need to have CVSS scores from the vendor (if available) and NVD (if available), no more.
+However, it is possible this requirement changes in the future, and we may want to support more than just two
+sources for vulnerability data.
+
+The focus is on CVEs and RHSAs, RHEAs, and RHBAs. This document will not consider other types of advisories such as
+Ubuntu's USNs.
+
+The current API looks like the following:
+
+```
+message VulnerabilityReport {
+  message Vulnerability {
+    enum Severity {
+      SEVERITY_UNSPECIFIED = 0;
+      SEVERITY_LOW = 1;
+      SEVERITY_MODERATE = 2;
+      SEVERITY_IMPORTANT = 3;
+      SEVERITY_CRITICAL = 4;
+    }
+    message CVSS {
+      message V2 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      message V3 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      V2 v2 = 1;
+      V3 v3 = 2;
+    }
+    ...
+    string severity = 6;
+    ...
+    CVSS cvss = 12;
+  }
+  ...
+}
+```
+
+## Decision
+
+The VulnerabilityReport API will be extended as follows:
+
+```
+message VulnerabilityReport {
+  message Vulnerability {
+    enum Severity {
+      SEVERITY_UNSPECIFIED = 0;
+      SEVERITY_LOW = 1;
+      SEVERITY_MODERATE = 2;
+      SEVERITY_IMPORTANT = 3;
+      SEVERITY_CRITICAL = 4;
+    }
+    message CVSS {
+      message V2 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      message V3 {
+        float base_score = 1;
+        string vector = 2;
+      }
+      V2 v2 = 1;
+      V3 v3 = 2;
+      string updater = 3; <-- New field.
+      string cvss_url = 4; <-- New field
+    }
+    ...
+    string severity = 6;
+    ...
+    CVSS cvss = 12;
+    repeated CVSS cvss_metrics = 13; <-- New field.
+  }
+  ...
+}
+```
+
+There will be a new type plus two new fields added:
```

Comment:
Thoughts on expanding this to include how the new fields/scores will be represented in Central's `storage.ImageScan` object? 

If this ADR is meant to describe only the changes specific to Scanner V4, so be it.  However, understanding how the new fields here get translated into the fields in `storage.ImageName` and then displayed in the UI (per the in-progress UI/UX design) would help determine if and where we may need to change things (like the updater name) to have a complete implementation.

Reply:
Not sure if this information helps but this ADR is intended to finalize the changes on the Scanner side. It does not cover core workflow and UX decisions.

I am not entirely certain how the new fields will translate into the fields in storage.ImageName. But we have agreed on these new fields based on multiple meetings with the UX and core workflow team. The UX team is aware of the new fields and has agreed that they align with their new design.

Does that address your concern?

Reply:
> Does that address your concern?

It does not.  We may not be able to accurately define change to one component (Scanner in this case) of the overall architecture without understanding what the responsibilities/contracts will be for each component. 

For example, it seems that in Central we'll be displaying a table for each vuln that will look similar to:

CVSS | Score Breakdown
--- | ---
9.7 (Red Hat V3) | CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:H
7.6 (NVD V2) | CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:H
7.4 (OSV V1) | CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:H

So 'something' will need to to translate the `updater` name in the vulnerability report to `Red Hat`, `NVD`, `OSV`, etc..  Looking at the JSON example, the updater names are taken verbatim from scanner / ClairCore internals and will be exposed via the API:
```
alpine-community-v3.13-updater
alpine-main-v3.4-updater
aws-AL2023-updater
debian/updater
stackrox-manual
oracle-2008-updater
RHEL8-rhel-8.1-e4s
RHEL7-ansible-2.9
RHEL6-amq-clients-2-including-unpatched
RHEL6-rhsso-including-unpatched
suse-updater-suse.linux.enterprise.server.11
etc.
```

How will Central map those ^^^ updater names to the nice `Red Hat`, `NVD`, `OSV`, etc. names?

Instead of exposing the internal updater name should we instead expose some other 'key' that can be mapped more cleanly - so that the internals of ClairCore/matcher do not bleed into the rest of ACS and allow them to be changed independently?

Reply:
I made a proposal to convert the updater names to enums
https://github.com/stackrox/stackrox/pull/10721/files#r1705734775

Reply:
FWIW, the updater name was just one example of 'something that would be realized' by walking through a use case end to end.  Same concern for the urls, the vectors, etc. 

Hypothetically: would a 'major database overhaul' be needed to store this data, or 'major UI redesign' be needed to display this data that could be avoided by changing a datatype in this API? or separating out the vector into individual fields? etc. (I don't believe so, but by going through the exercise we can be confident this is the most efficient way to represent this data and has minimal impact to other areas of ACS given known use cases - up to you)

Reply:
That makes sense we might have to update `storage.EmbeddedVulnerability` since we set the scores in [convert.go](https://github.com/stackrox/stackrox/blob/e7889e10fb1775d5a7e83d2ad771843366d99788/pkg/scanners/scannerv4/convert.go#L172)
What do you think? @RTann

Reply:
I believe `storage.EmbeddedVulnerability` will need to be extended, but it's been a long time since I've worked closely on how Central stores things (pre PostgreSQL days at StackRox). As Yi mentioned, we are working closely with UX and Core Workflows, and I believe this was all discussed in the previous meetings.

@ksurabhi91 do you have ideas for what changes to the `storage` protos would be needed and how invasive they'd be?

Reply:
Core workflow team has agreed that they will make the Central changes after the vulnerability protos change by the Scanner team. So I think we should be good for this ADR.